src/tequila/apps/qcc/qcc.py

```diff
@@ -0,0 +1,394 @@
+            for idx in range(len(self.generators[-1]))
+        }
+
+        for idx in range(len(self.generators[-1]) - 1, -1, -1):
+            gen_binary = generators_binary["g{}".format(idx)]
+            gen_x = gen_binary.binary[: self.n_qubits]
+            gen_z = gen_binary.binary[self.n_qubits :]
+            angle = self.angles[-1]["t{}".format(idx)]
+            introduced_terms = []
+            for h_idx in range(len(self.binary_repr)):
+                h_term = self.binary_repr[h_idx]
+                h_term_x = h_term.binary[: self.n_qubits]
+                h_term_z = h_term.binary[self.n_qubits :]
+                h_coeff = h_term.coeff
+                if (np.dot(h_term_x, gen_z) + np.dot(h_term_z, gen_x)) % 2 == 1:  # non-commutative
+                    self.binary_repr[h_idx] = BinaryPauli(h_coeff * np.cos(angle), h_term.binary)
+                    phase = binary_phase(gen_binary.binary, h_term.binary, self.n_qubits)
+                    product_binary = (gen_binary.binary + h_term.binary) % 2
+                    new_coeff = 1j * h_coeff * phase * np.sin(angle)
+                    introduced_terms.append(BinaryPauli(new_coeff, product_binary))
+            self.binary_repr += introduced_terms
+
+        # update self.hamiltonian from self.binary_repr
+        binary_h = BinaryHamiltonian(
+            [
+                BinaryPauliString(binary_pauli.binary, binary_pauli.coeff)
+                for binary_pauli in self.binary_repr
+            ]
+        )
+        self.hamiltonian = binary_h.to_qubit_hamiltonian()
+        self.hamiltonian = self.hamiltonian.simplify(threshold=compression_threshold)
+        # reassigning `self.binary_repr` seems redundant here, but it lets us consolidate identical binary strings
+        # without explicitly performing a search over the binary terms earlier:
+        self.binary_repr = [
+            pauli.binary(n_qubits=self.n_qubits) for pauli in self.hamiltonian.paulistrings
+        ]
+        n_terms = len(self.hamiltonian)
+        print("Number of terms in iQCC Hamiltonian: {}".format(n_terms))
+        self.n_terms.append(n_terms)
+        return self.hamiltonian
+
+    def grad_norm(self, order=1):
+        """
+        Computes the `order`-norm of the gradients. Here, the "gradient" directions denote
+        the non-zero gradient partitions of the Pauli group. Hence, each partition contributes
+        a single gradient component to the gradient norm.
+
+        Arguments:
+            order (int): positive integer specifying the order of norm to be used.
+        Returns:
+            float: the computed gradient norm.
+        """
+        if not hasattr(self, "grad_partitions"):
+            raise AttributeError("No previously computed gradient partition to check norm for!")
+
+        return sum([partition[1] ** order for partition in self.grad_partitions]) ** (1 / order)
+
+    def do_qcc(self, n_gen=1, method="COBYLA", grad_threshold=1e-6, x0=None):

```

Comment:
method should be an internal parameter (same as in the adapt module).
See commend above for the AdaptParameters integration.
You would use it internally then as this:

result = tq.minimize(...., **self.parameters.optimizer_arguments)

the default of optimizer_arguments is:
optimizer_arguments = {"method":"BFGS"} 

this allows users to control the optimizer more specifically:

e.g. 

solver = QCC(optimizer_arguments={"method":"cobyla", silent:True, "method_options"={...}}